docs(maven-rate-limit): add documentation about configuration maven caching repository (#1428)
codefresh-io/docs.codefresh.io: _docs/kb/articles/maven-rate-limit.md

---
title: "Maven Central Repository rate limiting in Pipelines"
description: 
group: kb
sub-group: articles
toc: true
kb: true
ht: false
common: false
categories: [Pipelines]
support-reviewed: 2023-04-18 LG
---

## Overview

The build failed with following error during installation of maven dependencies:
```shell
Exception in thread "main" java.io.IOException: 
Server returned HTTP response code: 429 for URL: 
https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.8.6/apache-maven-3.8.6-bin.zip
```

## Details

> Maven Central (operated by Sonatype) rate-limits traffic per IP, and the combined volume from all customers running on shared runtimes can trigger that limit — causing your builds to fail with HTTP 429 Too Many Requests.

Reviewer: also add in parentheses "(when facing the rate limit for a long time it might result in 403)"

> Codefresh SaaS pipelines share a single egress IP. Maven Central (operated by Sonatype) rate-limits traffic per IP, and the combined volume from all customers running on shared runtimes can trigger that limit — causing your builds to fail with HTTP 429 Too Many Requests.

Reviewer: ```suggestion
Maven Central (operated by Sonatype) rate-limits traffic per IP, and the combined volume from all customers running on shared runtimes can trigger that limit — causing your builds to fail with HTTP 429 Too Many Requests.
```

> This issue only affects pipelines running on Codefresh shared SaaS runtimes:

Reviewer: ```suggestion
This issue mostly affects pipelines running on Codefresh shared SaaS runtimes:
```

Reviewer: "....which is `Cloud Builds` or these runtimes:"

    system/linux_paying_plan
    system/linux_non_paying_plan
    system/codefresh-enterprise

>     system/plan/linux

Reviewer: Do our customers know about these runtimes?

> If your pipelines run on a hybrid runner (your own Kubernetes cluster), your builds already use your own egress IP and are not affected by this issue.

Reviewer: ```suggestion
If your pipelines run on a hybrid runner (your own Kubernetes cluster), your builds already use your own egress IP and you might not face this issue. However, if it happens this guide is still relevant for you.
```

> The fix is straightforward: route your Maven dependency downloads through a caching proxy that sits between your pipeline and Maven Central. After the first download of each artifact, all subsequent requests are served from the cache — Maven Central never sees repeated traffic from your builds.

Reviewer: ```suggestion
**The fix is straightforward**: route your Maven dependency downloads through a caching proxy that sits between your pipeline and Maven Central. After the first download of each artifact, all subsequent requests are served from the cache — Maven Central never sees repeated traffic from your builds.

There are a **couple of options** to proceed with:
- AWS CodeArtifact
- Google Artifact Registry
- Azure Artifacts
- Self-hosted Nexus Repository Manager

```

      <!-- Redirect all external Maven traffic through the proxy -->
      <mirrorOf>external:*</mirrorOf>
    </mirror>
  </mirrors>

</settings>
```

### Step 4: Add a token refresh step to your Codefresh pipeline

CodeArtifact auth tokens expire every 12 hours, so you need to fetch a fresh one at the start of each build. Add this as the first step in your pipeline:
```yaml
version: "1.0"

steps:
  clone:
    title: "Cloning repository"
    type: "git-clone"
    repo: "${{CF_REPO_OWNER}}/${{CF_REPO_NAME}}"
    revision: "${{CF_BRANCH}}"
    git: "github"

  get_codeartifact_token:
    title: Refresh CodeArtifact auth token
    image: amazon/aws-cli
    working_directory: "${{clone}}"
    commands:
      - echo CODEARTIFACT_AUTH_TOKEN=$(aws codeartifact get-authorization-token --domain codefresh-org --domain-owner ${DOMAIN_OWNER} --region us-east-1 --query authorizationToken --output text) >> ${{CF_VOLUME_PATH}}/env_vars_to_export 

  build:
    title: Maven build
    image: maven:3.9-eclipse-temurin-21
    working_directory: "${{clone}}"
    commands:
      - mvn -s ci/settings.xml clean install

> ```

Reviewer: Templated variables like {{var}} become empty in the final doc -- please escape them. See the dollar signs on the picture:

<img width="332" height="175" alt="Image" src="https://github.com/user-attachments/assets/7ef41656-ee09-407a-9d85-43ce093a3588" />

          https://maven.apache.org/xsd/settings-1.0.0.xsd">

  <servers>
    <server>
      <id>artifact-registry</id>
      <username>_json_key</username>
      <password>${env.GOOGLE_JSON_KEY_FILE}</password>
    </server>
  </servers>
          
  <mirrors>
    <mirror>
      <id>artifact-registry</id>
      <name>GCP Artifact Registry — Maven Central Proxy</name>
      <url>https://LOCATION-maven.pkg.dev/PROJECT_ID/REPOSITORY_NAME/</url>
      <mirrorOf>central</mirrorOf>
    </mirror>
  </mirrors>

</settings>
```
### Step 4: Authenticate in Codefresh pipeline

Store your service account json key file as a Codefresh secret variable named `GOOGLE_JSON_KEY_FILE`. Maven picks it
up automatically via the `${env.GOOGLE_JSON_KEY_FILE}` reference in `settings.xml`.

```yaml
version: "1.0"

steps:
  clone:
    title: "Cloning repository"
    type: "git-clone"
    repo: "${{CF_REPO_OWNER}}/${{CF_REPO_NAME}}"
    revision: "${{CF_BRANCH}}"
    git: "github"

  build:
    title: Maven build
    image: maven:3.9-eclipse-temurin-21
    working_directory: "${{clone}}"
    commands:
      - mvn -s ci/settings.xml clean install

> ```

Reviewer: same issue with templating here

          xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0
          https://maven.apache.org/xsd/settings-1.0.0.xsd">

  <servers>
    <server>
      <id>azure-artifacts</id>
      <username>AzureDevOps</username>
      <password>${env.AZURE_ARTIFACTS_PAT}</password>
    </server>
  </servers>

  <mirrors>
    <mirror>
      <id>azure-artifacts</id>
      <name>Azure Artifacts — Maven Central Proxy</name>
      <url>https://pkgs.dev.azure.com/MY_ORG/MY_PROJECT/_packaging/maven-proxy/maven/v1</url>
      <mirrorOf>external:*</mirrorOf>
    </mirror>
  </mirrors>

</settings>
```

### Step 5: Set the PAT in your Codefresh pipeline

Store your PAT as a Codefresh secret variable named `AZURE_ARTIFACTS_PAT`. Maven picks it
up automatically via the `${env.AZURE_ARTIFACTS_PAT}` reference in `settings.xml`.

```yaml
version: "1.0"

steps:
  clone:
    title: "Cloning repository"
    type: "git-clone"
    repo: "${{CF_REPO_OWNER}}/${{CF_REPO_NAME}}"
    revision: "${{CF_BRANCH}}"
    git: "github"

  build:
    title: Maven build
    image: maven:3.9-eclipse-temurin-21
    working_directory: "${{clone}}"
    commands:
      - mvn -s ci/settings.xml clean install

> ```

Reviewer: same issue with templating

## Configure Maven Proxy Cache using Self-Hosted Nexus Repository Manager

Nexus OSS is free. Infrastructure cost depends on your setup — a minimal deployment (2 vCPU, 4 GB RAM, 100 GB storage) runs roughly $60–120/month on most cloud providers. A production-grade setup with SSD storage and egress is closer to $200–500/month.

>_**Note**:_
>
>The reason the YAML editor does not catch this at this time is because it istechnically a valid YAML file.

> 

Reviewer: ```suggestion
```

  <servers>
    <server>
      <id>nexus</id>
      <username>${env.NEXUS_USER}</username>
      <!-- Token is injected at build time via environment variable -->
      <password>${env.NEXUS_PASSWORD}</password>
    </server>
  </servers>

  <mirrors>
    <mirror>
      <id>nexus</id>
      <name>Nexus Maven Central Proxy</name>
      <url>https://nexus.your-domain.com/repository/maven-central-proxy/</url>
      <!-- Redirect all external Maven traffic through the proxy -->
      <mirrorOf>external:*</mirrorOf>
    </mirror>
  </mirrors>

</settings>
```

### Step 4: Add credentials to your Codefresh pipeline

Add Codefresh secret variables named `NEXUS_USER` and 'NEXUS_PASSWORD''. Maven picks it
up automatically via the `${env.NEXUS_USER}` and `${env.NEXUS_PASSWORD}` reference in `settings.xml`.:

```yaml
version: "1.0"

steps:
  clone:
    title: "Cloning repository"
    type: "git-clone"
    repo: "${{CF_REPO_OWNER}}/${{CF_REPO_NAME}}"
    revision: "${{CF_BRANCH}}"
    git: "github"

  build:
    title: Maven build
    image: maven:3.9-eclipse-temurin-21
    working_directory: "${{clone}}"
    commands:
      - mvn -s ci/settings.xml clean install

> ```

Reviewer: same issue with templating

## Verifying the Proxy Works

> Check logs for the step where maven dependencies installations are executed

Reviewer: ```suggestion
Check the build logs for the pipeline step where maven dependencies installation is executed -- if you see your proxy URL instead of repo.maven.apache.org, the caching proxy is working as expected:
```

```
Downloading from proxy: https://your-proxy-url/org/apache/...
```

> If you see your proxy URL instead of repo.maven.apache.org, the redirect is working.

Reviewer: ```suggestion
```
